packages/keystone/adapterCache.js

```diff
@@ -0,0 +1,335 @@
+const { getLogger } = require('./logging')
+
+const UPDATED_AT_FIELD = 'updatedAt'
+const STATE_REDIS_KEY_PREFIX = 'adapterCacheState'
+
+const logger = getLogger('adapterCache')
+
+
+class AdapterCacheMiddleware {
+
+    constructor (config) {
+        try {
+            const parsedConfig = JSON.parse(config)
+
+            this.enabled = !!get(parsedConfig, 'enable', false)
+
+            // Cache: tableName -> queryKey -> { response, lastUpdate }
+            this.cache = {}
+
+            // Redis is used as State:
+            // State: table_name -> lastUpdate
+            this.redisClient = this.getStateRedisClient()
+
+            // This mechanism allows to skip caching some lists.
+            // Useful for hotfixes or disabling cache for business critical lists
+            //
+            // includeAllLists is Bool:
+            // If includeAllLists is True:
+            // ==> "includedLists" is ignored.
+            // ==> all lists, that are not in "excludedLists" are cached.
+            // Else:
+            // ==> only lists, that are in "includedLists" are cached.
+            // ==> lists that are in "excludedLists" are NOT cached.
+            this.includeAllLists = !!get(parsedConfig, 'includeAllLists', false)
+            this.includedLists = get(parsedConfig, 'includedLists', [])
+            this.excludedLists = get(parsedConfig, 'excludedLists', [])
+
+            // Logging allows to get the percentage of cache hits
+            this.logging = get(parsedConfig, 'logging', false)
+            this.totalRequests = 0
+            this.cacheHits = 0
+
+            // Debug mode allows to get full history of operations with cache including cache dumps.
+            // Useful for local debugging
+            // You shouldn't allow this on production as it will lead to memory leak!
+            this.debugMode = !!get(parsedConfig, 'debug', false)
+            this.cacheHistory = {}
+            this.cacheCallHistory = []
+            if (this.debugMode) {
+                logger.warn(
+                    'ADAPTER CACHE HAS DEBUG MODE TURNED ON. THIS WILL LEAD TO MEMORY LEAK ERRORS IN NON LOCAL ENVIRONMENT')
+            }
+        } catch (e) {
+            this.enabled = false
+            logger.warn(`ADAPTER_CACHE: Bad config! reason ${e}`)
+        }
+    }
+
+    /**
+     * Sometimes we might get a situation when two instances try to update state with different timestamp values.
+     * So, we might have two commands ordered to execute by internal redis queue:
+     * 1. SET some_key 18:00
+     * 2. SET some_key 17:59
+     * In result (GET some_key) will equal 17:59, but the correct value is 18:00!
+     * So to counter this we write custom redis function that will update value only if it is bigger!
+     */
+    getStateRedisClient () {
+        const updateTimeStampFunction = {
+            numberOfKeys: 1,
+            lua: `
+            local time = tonumber(ARGV[1])
```

Comment:
is it atomic?

Reply:
No, since it has ` tonumber(redis.call('GET', KEYS[1]))` instruction inside it.